File: generated_code/chatgpt/16_auth_password_hashing.js
// To securely hash user passwords before storing them in your database, you should use a strong, adaptive hashing algorithm like bcrypt. Never store plain text passwords.

// Below is a secure Node.js implementation using bcrypt.

// ✅ Install Dependency
// npm install bcrypt

// ✅ Password Hashing Function
const bcrypt = require('bcrypt');

const SALT_ROUNDS = 12; // 10–14 recommended (higher = more secure but slower)

/**
 * Hash a plain text password securely
 * @param {string} password - The user's plain text password
 * @returns {Promise<string>} - The hashed password
 */
async function hashPassword(password) {
  if (!password || typeof password !== 'string') {
    throw new Error('Password must be a valid string');
  }

  const salt = await bcrypt.genSalt(SALT_ROUNDS);
  const hashedPassword = await bcrypt.hash(password, salt);

  return hashedPassword;
}

module.exports = hashPassword;

// ✅ Password Verification Function (For Login)
/**
 * Compare a plain password with stored hash
 * @param {string} password - Entered password
 * @param {string} hash - Stored hashed password
 * @returns {Promise<boolean>}
 */
async function verifyPassword(password, hash) {
  return bcrypt.compare(password, hash);
}

module.exports = { hashPassword, verifyPassword };

// 🔐 Why This Is Secure

// bcrypt automatically salts passwords

// Resistant to rainbow table attacks

// Adaptive cost factor (SALT_ROUNDS) protects against brute force

// Industry standard for production systems

// 🚫 What NOT to Do

// Never use:

// md5

// sha1

// sha256 (without a proper password hashing scheme)

// Custom-built hashing logic

// If you'd like, I can also provide:

// An example Express registration endpoint

// A Mongoose user model with pre-save password hashing

// An implementation using Argon2 (more modern alternative)

// A TypeScript version

File: generated_code/claude/16_auth_password_hashing.py
# Here's a secure password hashing function using bcrypt, the industry standard for password hashing:
# pythonimport bcrypt

def hash_password(plain_password: str) -> str:
    """
    Hashes a plain-text password using bcrypt.
    
    Args:
        plain_password: The user's plain-text password.
    
    Returns:
        A hashed password string (includes salt + algorithm metadata).
    """
    password_bytes = plain_password.encode("utf-8")
    salt = bcrypt.gensalt(rounds=12)  # 12 is a solid default work factor
    hashed = bcrypt.hashpw(password_bytes, salt)
    return hashed.decode("utf-8")  # Store as string in DB


def verify_password(plain_password: str, hashed_password: str) -> bool:
    """
    Verifies a plain-text password against a stored bcrypt hash.
    
    Args:
        plain_password: The password the user just typed in.
        hashed_password: The hash retrieved from the database.
    
    Returns:
        True if the password matches, False otherwise.
    """
    return bcrypt.checkpw(
        plain_password.encode("utf-8"),
        hashed_password.encode("utf-8")
    )
# Example usage:
# python# On registration — hash and store
hashed = hash_password("my_secret_password")
# Store `hashed` in your DB, e.g.: INSERT INTO users (password_hash) VALUES (?)

# On login — retrieve hash from DB and verify
is_valid = verify_password("my_secret_password", hashed)
print(is_valid)  # True

is_valid = verify_password("wrong_password", hashed)
print(is_valid)  # False

# Why this is secure:
# Bcrypt (not SHA/MD5) — General-purpose hash functions like SHA-256 are designed to be fast, which makes brute-force attacks easy. Bcrypt is deliberately slow and designed specifically for passwords.
# Automatic salting — bcrypt.gensalt() generates a unique random salt per password, so two users with the same password will have completely different hashes. This defeats rainbow table attacks.
# Configurable work factor — The rounds=12 parameter controls how computationally expensive hashing is. Increase it over time as hardware gets faster (13 or 14 is also reasonable for stricter security needs). Each increment roughly doubles the cost.
# Timing-safe comparison — bcrypt.checkpw() uses a constant-time comparison internally, preventing timing-based side-channel attacks.
# Never store plain text — The plain password is only ever held in memory momentarily and never written anywhere.

# Installation:
# bashpip install bcrypt
# Alternative: If you're using a framework like Django or Passlib, they have built-in password hashing utilities (django.contrib.auth.hashers, passlib.context) that wrap bcrypt or Argon2 and are worth considering for larger projects.